### PR TITLE
7466 PSM upgradability

### LINKED ISSUE #7466: make the PSM contract upgradable
## What is the Problem Being Solved?

Must be upgradable for vaults release. Note that we won't be upgrading the instance of the PSM in Pismo release because we're reinstantiating in the bulldozer (https://github.com/Agoric/agoric-sdk/issues/6645). But the new instance must be upgradable going forward.

### PR DESCRIPTION
**Pull request by turadg**
closes: #7466

## Description

Makes the PSM contract upgradable.

I tried to keep the commits clean to show the code movements. I recommend first studying the changes by commit. But then because some code will have moved again, making review comments on the full *Files changed* view.

### Security Considerations

Authoritative data from db.

### Scaling Considerations

Can now be paged out

### Documentation Considerations

One small breaking change that `getMetrics()` returns a regular Subscriber instead of StoredSubscriber. The latter was never available to end users so okay.

### Testing Considerations

New swingset test

### CODE REVIEW

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+          governed: {
+            terms: psmTerms,
+            issuerKeywordRecord: { AUSD: anchor.issuer },
+            label: 'psmV1',
+          },
+        }),
+      );
+      trace('got governorTerms', governorTerms);
+
+      // Complete round-trip without upgrade
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js]: I think buildV1 is part of a complete round-trip, but this comment doesn't mark a location that's relevant for that.

File: packages/inter-protocol/src/psm/psm.js [hunk 2]
```diff
@@ -53,6 +61,36 @@ const { Fail } = assert;
 
 /** @typedef {import('@agoric/vat-data').Baggage} Baggage */
 
+export const customTermsShape = {
```
Comment (dckc) [inline packages/inter-protocol/src/psm/psm.js:70]: :clap:

File: packages/inter-protocol/src/psm/psm.js [hunk 3]
```diff
@@ -53,6 +59,36 @@ const { Fail } = assert;
...
+      type: ParamTypes.RATIO,
+      value: RatioShape,
+    },
+    MintLimit: { type: ParamTypes.AMOUNT, value: AmountShape },
+  },
+};
+harden(customTermsShape);
+
+export const privateArgsShape = {
+  feeMintAccess: M.any(),
```
Comment (dckc) [inline packages/inter-protocol/src/psm/psm.js]: Since its identity matters, I think `M.remotable()` would exclude promises.

File: packages/inter-protocol/test/swingsetTests/fluxAggregator/bootstrap-fluxAggregator-service-upgrade.js [hunk 4]
```diff
@@ -13,7 +13,7 @@ import { E } from '@endo/eventual-send';
 import { Far } from '@endo/marshal';
 import { makePromiseKit } from '@endo/promise-kit';
 
-const trace = makeTracer('BootWFUpg');
+const trace = makeTracer('BootFAUpg');
```
Comment (dckc) [inline packages/inter-protocol/test/swingsetTests/fluxAggregator/bootstrap-fluxAggregator-service-upgrade.js:16]: odd drive-by for this commit

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+     * }} vats
+     * @param {*} devices
+     */
+    bootstrap: async (vats, devices) => {
+      vatAdmin = await E(vats.vatAdmin).createVatAdminService(devices.vatAdmin);
+      ({ feeMintAccess, zoeService } = await E(vats.zoe).buildZoe(
+        vatAdmin,
+        undefined,
+        'zcf',
+      ));
```
Comment (dckc) [inline packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js:121]: Using promises would make clear that it's a one-time assignment:

```suggestion
      const { feeMintAccess, zoeService } = await E(vats.zoe).buildZoe(
        vatAdmin,
        undefined,
        'zcf',
      );
      bootPK.resolve({ feeMintAccess, zoeService });
```

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+import { PaymentPKeywordRecordShape } from '@agoric/zoe/src/typeGuards.js';
+import buildManualTimer from '@agoric/zoe/tools/manualTimer.js';
+import { E } from '@endo/eventual-send';
+import { Far } from '@endo/marshal';
+import { makePromiseKit } from '@endo/promise-kit';
+import { withAmountUtils } from '../../supports.js';
+
+const trace = makeTracer('BootPSMUpg');
+
+export const faV1BundleName = 'psmV1';
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js]: s/fa/psm/g. 

you had me confused for a while. I was trying to figure out why it was working with the fluxAggregator.  Particularly since the minor edit to a fluxAggregator test was in the same commit, so I found it first when searching for `faV1BundleName`!

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/test-psm-upgrade.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
+import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava.js';
+
+import { assert } from '@agoric/assert';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { resolve as importMetaResolve } from 'import-meta-resolve';
+
+import { buildVatController } from '@agoric/swingset-vat';
+import { faV1BundleName } from './bootstrap-psm-upgrade.js';
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/swingsetTests/psmUpgrade/test-psm-upgrade.js]: please rename to `psmV1BundleName`.

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+        type: ParamTypes.RATIO,
+        value: makeRatio(1n, minted.brand, 1n),
+      };
+      psmTerms.governedParams.GiveMintedFee = {
+        type: ParamTypes.RATIO,
+        value: makeRatio(1n, minted.brand, 1n),
+      };
+      psmTerms.governedParams.MintLimit = {
+        type: ParamTypes.AMOUNT,
+        value: AmountMath.make(minted.brand, scale6(1_000_000)),
```
Comment (dckc) [inline packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js]: slightly odd to not use `minted.make` defined a few lines up

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/test-psm-upgrade.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+test('PSM service upgrade', async t => {
+  /** @type {SwingSetConfig} */
+  const config = {
+    includeDevDependencies: true, // test's bootstrap has some deps not needed in production
+    defaultManagerType: 'local',
+    bundleCachePath: 'bundles/',
+    bootstrap: 'bootstrap',
+    vats: {
+      bootstrap: {
+        // TODO refactor to use bootstrap-relay.js
```
Comment (Chris-Hibbert) [inline packages/inter-protocol/test/swingsetTests/psmUpgrade/test-psm-upgrade.js]: s/TODO/XXX/

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/test-psm-upgrade.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { resolve as importMetaResolve } from 'import-meta-resolve';
+
+import { buildVatController } from '@agoric/swingset-vat';
+import { faV1BundleName } from './bootstrap-psm-upgrade.js';
+
+// so paths can be expresssed relative to this file and made absolute
+const bfile = name => new URL(name, import.meta.url).pathname;
+
+test('PSM service upgrade', async t => {
```
Comment (dckc) [inline packages/inter-protocol/test/swingsetTests/psmUpgrade/test-psm-upgrade.js:13]: These upgrade tests are so verbose that it's a little hard to see the forest for the trees. But it's not obvious how to do much better.

To confirm: this tests that before and after a null upgrade, we can do PSM trades, yes? If there's a difference between `testFunctionality1` and `testFunctionality2`, I don't see it.

File: packages/inter-protocol/src/psm/psm.js [hunk 11]
```diff
@@ -18,14 +18,15 @@ import {
...
+  provideAll,
+  provideEmptySeat,
 } from '@agoric/zoe/src/contractSupport/index.js';
 import {
+  AmountKeywordRecordShape,
   InstanceHandleShape,
   InvitationShape,
 } from '@agoric/zoe/src/typeGuards.js';
 import { E } from '@endo/eventual-send';
-import { Far } from '@endo/marshal';
```
Comment (dckc) [inline packages/inter-protocol/src/psm/psm.js:22]: :clap:

File: packages/inter-protocol/src/psm/psm.js [hunk 12]
```diff
@@ -105,10 +106,9 @@ export const prepare = async (zcf, privateArgs, baggage) => {
   const { anchorBrand, anchorPerMinted } = zcf.getTerms();
   console.log('PSM Starting', anchorBrand, anchorPerMinted);
 
-  const stableMint = await zcf.registerFeeMint(
-    'Minted',
-    privateArgs.feeMintAccess,
-  );
+  const { stableMint } = await provideAll(baggage, {
+    stableMint: () => zcf.registerFeeMint('Minted', privateArgs.feeMintAccess),
```
Comment (dckc) [inline packages/inter-protocol/src/psm/psm.js:122]: seems a little odd that we have to preserve the identity of the stable mint on our end; seems like calling `registerFeeMint` in each incarnation should be fine.

I suppose what we need to do is avoid registering the keyword `Minted` twice.

File: packages/inter-protocol/src/psm/psm.js [hunk 13]
```diff
@@ -212,7 +218,10 @@ export const prepare = async (zcf, privateArgs, baggage) => {
     // one immediately below. This `burnMinted`
     // happen only if the `atomicRearrange` does *not* throw.
     burnMinted(afterFee);
-    totalAnchorProvided = AmountMath.add(totalAnchorProvided, maxAnchor);
+    baggage.set(
+      'totalAnchorProvided',
+      AmountMath.add(baggage.get('totalAnchorProvided'), maxAnchor),
```
Comment (dckc) [inline packages/inter-protocol/src/psm/psm.js:282]: I can see why you were asking about record-like access to baggage. It's tempting to use a proxy. But I guess that's what exo does.

File: packages/inter-protocol/src/psm/psm.js [hunk 14]
```diff
@@ -149,13 +155,23 @@ export const prepare = async (zcf, privateArgs, baggage) => {
...
+        ),
+      ),
+  });
+  const topics = harden({
+    metrics: makeRecorderTopic('PSM metrics', metricsKit),
+  });
+
   const updateMetrics = () => {
-    metricsPublisher.publish(
+    E(metricsKit.recorder).write(
```
Comment (dckc) [inline packages/inter-protocol/src/psm/psm.js]: `void`?

File: packages/inter-protocol/src/psm/psm.js [hunk 15]
```diff
@@ -292,17 +308,23 @@ export const prepare = async (zcf, privateArgs, baggage) => {
     M.interface('PSM', {
       getMetrics: M.call().returns(M.remotable('MetricsSubscriber')),
       getPoolBalance: M.call().returns(anchorAmountShape),
+      getPublicTopics: M.call().returns({
+        metrics: M.any(),
```
Comment (dckc) [inline packages/inter-protocol/src/psm/psm.js]: `M.record()` maybe?

**Review by Chris-Hibbert — APPROVED** [review]
looks fine. requests for minor renaming/cleanup, but nothing that affects correctness.

File: packages/inter-protocol/src/psm/psm.js [hunk 16]
```diff
@@ -97,38 +139,45 @@ export const start = async (zcf, privateArgs, baggage) => {
...
-  const anchorPool = provideEmptyZcfSeat('anchorPoolSeat');
-  const feePool = provideEmptyZcfSeat('feePoolSeat');
-  const stage = provideEmptyZcfSeat('stageSeat');
+  const anchorPool = provideEmptySeat(zcf, baggage, 'anchorPoolSeat');
+  const feePool = provideEmptySeat(zcf, baggage, 'feePoolSeat');
+  const stage = provideEmptySeat(zcf, baggage, 'stageSeat');
 
-  let mintedPoolBalance = provide(baggage, 'mintedPoolBalance', () =>
+  // XXX access these through baggage directly so changes are saved
+  // TODO move into an Exo class state
```
Comment (dckc) [inline packages/inter-protocol/src/psm/psm.js]: TODO when? Not before landing this, right? Hm. I guess I can make my peace with that if you can.

**Review by dckc — APPROVED** [review]
raw baggage access... hm... I suppose I'm OK with it.

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/test-psm-upgrade.js (hunk 9 above)
Comment (dckc) [reply packages/inter-protocol/test/swingsetTests/psmUpgrade/test-psm-upgrade.js]: Would `bootstrap-relay.js` work here? I thought the custom bootstrap was required. I suspect `bootstrap-relay.js` would need some enhancement.

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js (hunk 6 above)
Comment (turadg) [reply packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js]: sorry bout that. I thought I rename those but maybe lost in git

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/test-psm-upgrade.js (hunk 9 above)
Comment (turadg) [reply packages/inter-protocol/test/swingsetTests/psmUpgrade/test-psm-upgrade.js]: what I had in mind is a test like the bootstrapTests that can use `EV` instead of `E` to make calls from an Ava context. but i'll just strike this comment. If I do get around to providing that, this test doesn't need to adopt that style and it's not tech debt to be in this style

File: packages/inter-protocol/src/psm/psm.js (hunk 3 above)
Comment (turadg) [reply packages/inter-protocol/src/psm/psm.js]: good call. I made a FeeMintAccessShape

File: packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js (hunk 8 above)
Comment (turadg) [reply packages/inter-protocol/test/swingsetTests/psmUpgrade/bootstrap-psm-upgrade.js]: yup, missed refactoring

File: packages/inter-protocol/src/psm/psm.js (hunk 12 above)
Comment (turadg) [reply packages/inter-protocol/src/psm/psm.js:122]: yup

File: packages/inter-protocol/src/psm/psm.js (hunk 15 above)
Comment (turadg) [reply packages/inter-protocol/src/psm/psm.js]: thanks for pointing out the any. there's actually a `TopicsRecordShape` already so I added that

File: packages/inter-protocol/src/psm/psm.js (hunk 16 above)
Comment (turadg) [reply packages/inter-protocol/src/psm/psm.js]: reworded
